### PR TITLE
[RISCV] Add TuneDisableLatencySchedHeuristic

### PR DESCRIPTION
**Pull request by wangpc-pp**
This tune feature will disable latency scheduling heuristic.

This can reduce the number of spills/reloads but will cause some
regressions on some cores.

CPU may add this tune feature if they find it's profitable.

### CODE REVIEW

**Comment by preames** [comment]
I suggest we let this sit for 2-3 days after the prior patch has landed.  As I said on the previous review, I think this is reasonable, but a) we had a bunch of discussion on this point and b) it'd be good to have some staging between the individual commits to simplify regression analysis.

**Comment by michaelmaitland** [comment]
For the recent scheduler patches, the common theme is we saw another target did something brought that functionality to RISC-V. How do we know that these changes are sensible defaults for RISC-V cores? Are you making measurements on any cores? Are they in order, out of order, both? In my experience tuning for different cores, there is often a difference between OOO and in order cores.

**Comment by mshockwave** [comment]
> This helps reduce register pressure for some cases.

Is it possible to provide some numbers to back this up? Preferably using some well known benchmarks like SPEC and/or llvm-test-suite

**Comment by wangpc-pp** [comment]
I added two experimental options: `-riscv-disable-latency-heuristic` and `-riscv-should-track-lane-masks` and evaluated the statistics (`regalloc.NumSpills`/`regalloc.NumReloads`) on llvm-test-suite (option: `-O3 -march=rva23u64`):
1. `-riscv-disable-latency-heuristic=true` and `-riscv-should-track-lane-masks=false`:
```
Program                                       regalloc.NumSpills                   regalloc.NumReloads                  
                                              00                 10       diff     00                  10       diff    
SingleSour...ce/UnitTests/matrix-types-spec    8823.00            6166.00 -2657.00 15603.00            13403.00 -2200.00
External/S...rate/510.parest_r/510.parest_r   43817.00           43262.00  -555.00 87058.00            87033.00   -25.00
External/S...017speed/625.x264_s/625.x264_s    2373.00            1991.00  -382.00  4808.00             4287.00  -521.00
External/S...2017rate/525.x264_r/525.x264_r    2373.00            1991.00  -382.00  4808.00             4287.00  -521.00
MultiSourc...ks/ASCI_Purple/SMG2000/smg2000    2684.00            2334.00  -350.00  4820.00             4349.00  -471.00
MultiSourc...nchmarks/FreeBench/pifft/pifft     442.00             126.00  -316.00   595.00              281.00  -314.00
MultiSourc.../Applications/JM/ldecod/ldecod    1335.00            1131.00  -204.00  2311.00             2142.00  -169.00
External/S...00.perlbench_s/600.perlbench_s    4354.00            4154.00  -200.00  9615.00             9435.00  -180.00
External/S...00.perlbench_r/500.perlbench_r    4354.00            4154.00  -200.00  9615.00             9435.00  -180.00
MultiSourc.../Applications/JM/lencod/lencod    3368.00            3172.00  -196.00  7261.00             7069.00  -192.00
External/S...te/538.imagick_r/538.imagick_r    4163.00            4000.00  -163.00 10354.00             9964.00  -390.00
MultiSourc...ch/consumer-lame/consumer-lame     722.00             559.00  -163.00  1098.00              994.00  -104.00
External/S...ed/638.imagick_s/638.imagick_s    4163.00            4000.00  -163.00 10354.00             9964.00  -390.00
MultiSource/Applications/oggenc/oggenc          970.00             817.00  -153.00  2327.00             2120.00  -207.00
MultiSourc...e/Applications/ClamAV/clamscan    2072.00            1937.00  -135.00  4836.00             4648.00  -188.00
      regalloc.NumSpills                            regalloc.NumReloads                           
run                   00            10         diff                  00            10         diff
mean   87.747460          84.068699    -3.678761     1371.475285         1357.146154  -3.792937   
```
2. `-riscv-disable-latency-heuristic=false` and `-riscv-should-track-lane-masks=true`:
```
Program                                       regalloc.NumSpills                 regalloc.NumReloads                 
                                              00                 01      diff    00                  01       diff   
SingleSour...ce/UnitTests/matrix-types-spec   8823.00            8233.00 -590.00 15603.00            15020.00 -583.00
MultiSourc...ch/consumer-lame/consumer-lame    722.00             689.00  -33.00  1098.00             1065.00  -33.00
MultiSourc...s/Prolangs-C/football/football    248.00             250.00    2.00   349.00              350.00    1.00
MultiSourc...ench/telecomm-gsm/telecomm-gsm    182.00             181.00   -1.00   196.00              195.00   -1.00
MultiSourc...Benchmarks/7zip/7zip-benchmark   1272.00            1273.00    1.00  2436.00             2437.00    1.00
MicroBench...arks/ImageProcessing/Blur/blur    114.00             113.00   -1.00   136.00              136.00    0.00
MultiSourc...rks/mediabench/gsm/toast/toast    182.00             181.00   -1.00   196.00              195.00   -1.00
MultiSourc...gs-C/TimberWolfMC/timberwolfmc   1196.00            1195.00   -1.00  2036.00             2029.00   -7.00
SingleSour.../execute/GCC-C-execute-pr36321      0.00               0.00    0.00                                 0.00
SingleSour.../execute/GCC-C-execute-pr36077      0.00               0.00    0.00                                 0.00
SingleSour...xecute/GCC-C-execute-pr33779-1      0.00               0.00    0.00                                 0.00
SingleSour.../execute/GCC-C-execute-pr33669      0.00               0.00    0.00                                 0.00
SingleSour.../execute/GCC-C-execute-pr33631      0.00               0.00    0.00                                 0.00
SingleSour.../execute/GCC-C-execute-pr33382      0.00               0.00    0.00                                 0.00
SingleSour.../execute/GCC-C-execute-pr37102      0.00               0.00    0.00                                 0.00
      regalloc.NumSpills                            regalloc.NumReloads                           
run                   00            01         diff                  00            01         diff
mean   87.747460          87.445573    -0.301887     1371.475285         1369.091255  -0.303338   
```
3. `-riscv-disable-latency-heuristic=true` and `-riscv-should-track-lane-masks=true`:
```
Program                                       regalloc.NumSpills                   regalloc.NumReloads                  
                                              00                 11       diff     00                  11       diff    
SingleSour...ce/UnitTests/matrix-types-spec    8823.00            6320.00 -2503.00 15603.00            13544.00 -2059.00
External/S...rate/510.parest_r/510.parest_r   43817.00           43262.00  -555.00 87058.00            87033.00   -25.00
External/S...017speed/625.x264_s/625.x264_s    2373.00            1991.00  -382.00  4808.00             4287.00  -521.00
External/S...2017rate/525.x264_r/525.x264_r    2373.00            1991.00  -382.00  4808.00             4287.00  -521.00
MultiSourc...ks/ASCI_Purple/SMG2000/smg2000    2684.00            2334.00  -350.00  4820.00             4349.00  -471.00
MultiSourc...nchmarks/FreeBench/pifft/pifft     442.00             126.00  -316.00   595.00              281.00  -314.00
MultiSourc.../Applications/JM/ldecod/ldecod    1335.00            1131.00  -204.00  2311.00             2142.00  -169.00
External/S...00.perlbench_s/600.perlbench_s    4354.00            4154.00  -200.00  9615.00             9435.00  -180.00
External/S...00.perlbench_r/500.perlbench_r    4354.00            4154.00  -200.00  9615.00             9435.00  -180.00
MultiSourc.../Applications/JM/lencod/lencod    3368.00            3172.00  -196.00  7261.00             7069.00  -192.00
External/S...te/538.imagick_r/538.imagick_r    4163.00            4000.00  -163.00 10354.00             9964.00  -390.00
MultiSourc...ch/consumer-lame/consumer-lame     722.00             559.00  -163.00  1098.00              994.00  -104.00
External/S...ed/638.imagick_s/638.imagick_s    4163.00            4000.00  -163.00 10354.00             9964.00  -390.00
MultiSource/Applications/oggenc/oggenc          970.00             817.00  -153.00  2327.00             2120.00  -207.00
MultiSourc...e/Applications/ClamAV/clamscan    2072.00            1937.00  -135.00  4836.00             4648.00  -188.00
      regalloc.NumSpills                            regalloc.NumReloads                           
run                   00            11         diff                  00            11         diff
mean   87.747460          84.142235    -3.605225     1371.475285         1357.692308  -3.724238   
```

We can see that both options can reduce the mean of spills/reloads. `ShouldTrackLaneMasks` has smaller influence because only vector registers (with sub-registers) can benefit from this.

I didn't run these tests on real hardwares, so these data may not be so convincing. I'd appreciate it if you can evaluate this on some platforms, that will be helpful. If you find this common setting is not suitable for your microarchitectures, please let me know, we can make it a tune feature. All I want is just to unify the common sched policy and make part of the policy being tune features.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll [hunk 1]
```diff
@@ -727,27 +728,49 @@ define void @buildvec_seq_v4i16_v2i32(ptr %x) {
...
-; CHECK-NEXT:    vse32.v v9, (a2)
-; CHECK-NEXT:    vse32.v v9, (a3)
-; CHECK-NEXT:    vse32.v v9, (a4)
-; CHECK-NEXT:    vmv.v.i v9, 0
-; CHECK-NEXT:    li a0, 1
-; CHECK-NEXT:    vslide1down.vx v9, v9, a0
-; CHECK-NEXT:    vse32.v v8, (a5)
-; CHECK-NEXT:    vse32.v v9, (a6)
-; CHECK-NEXT:    ret
+; RV32-LABEL: buildvec_vid_step1o2_v4i32:
```
Comment (mshockwave) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:731]: why do we have different scheduling between RV32 and RV64?

File: llvm/test/CodeGen/RISCV/rvv/vselect-vp.ll [hunk 2]
```diff
@@ -354,30 +354,51 @@ define <vscale x 32 x i32> @select_nxv32i32(<vscale x 32 x i1> %a, <vscale x 32
...
+; CHECK-NEXT:    slli a1, a1, 5
 ; CHECK-NEXT:    sub sp, sp, a1
-; CHECK-NEXT:    .cfi_escape 0x0f, 0x0d, 0x72, 0x00, 0x11, 0x10, 0x22, 0x11, 0x10, 0x92, 0xa2, 0x38, 0x00, 0x1e, 0x22 # sp + 16 + 16 * vlenb
+; CHECK-NEXT:    .cfi_escape 0x0f, 0x0d, 0x72, 0x00, 0x11, 0x10, 0x22, 0x11, 0x20, 0x92, 0xa2, 0x38, 0x00, 0x1e, 0x22 # sp + 16 + 32 * vlenb
 ; CHECK-NEXT:    csrr a1, vlenb
-; CHECK-NEXT:    slli a1, a1, 3
+; CHECK-NEXT:    slli a1, a1, 4
+; CHECK-NEXT:    add a1, sp, a1
+; CHECK-NEXT:    addi a1, a1, 16
+; CHECK-NEXT:    vs8r.v v16, (a1) # Unknown-size Folded Spill
```
Comment (mshockwave) [inline llvm/test/CodeGen/RISCV/rvv/vselect-vp.ll:364]: do we have more spills / reloads in this function?

**Review by mshockwave — APPROVED** [review]
LGTM . I do have some minor questions but they're not blocking.

**Comment by preames** [comment]
Given @michaelmaitland's data, @wangpc-pp  the burden shifts to you to clearly justify which cases this is profitable and figure out how to selectively enable only in profitable cases.  I agree with @michaelmaitland's conclusion that this should not move forward otherwise.  

@michaelmaitland Can you say anything about the magnitude of regression in either case?  I assume they were statistically significant given you mention them, but are these small regressions or largish ones?

**Comment by michaelmaitland** [comment]
> Given @michaelmaitland's data, @wangpc-pp the burden shifts to you to clearly justify which cases this is profitable and figure out how to selectively enable only in profitable cases. I agree with @michaelmaitland's conclusion that this should not move forward otherwise.
> 
> @michaelmaitland Can you say anything about the magnitude of regression in either case? I assume they were statistically significant given you mention them, but are these small regressions or largish ones?

* sjeng: 1.86% regression
* 557.xz_r: 1.14% regression
* 456.hmmer: 1.03% regression
* All other x in results were -1 < x < 1

**Comment by wangpc-pp** [comment]
Thanks for evaluating this! The data is very helpful! @michaelmaitland 

> Given @michaelmaitland's data, @wangpc-pp the burden shifts to you to clearly justify which cases this is profitable and figure out how to selectively enable only in profitable cases. I agree with @michaelmaitland's conclusion that this should not move forward otherwise.

I don't have other data other than the spill/reload data above. I don't know how to dynamically determine if a SchedDAG region will benefit from disabling it because we can only know `NumRegionInstrs` (we may change the function signature and pass DAG directly in the future so that we can analyse the region). AArch64 is the only target will disable it and almost all Apple's CPUs have this feature on (don't know if it is profitable or they are just some inertial copies when defining new processor).

Again, if the conclusion is that we shouldn't make it true by default, I can make it a tune feature. All I want is making scheduling infrastructure easy to tune for downstreams. :-)

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**